feat(chore): add support for fontawesome pro duotone icons

testinfected opened the pull request:
This PR adds support for FontAwesome Pro Duotone icons.

Duotone icons are composed of two maps instead of a single one, for ::before and ::after pseudo-elements.

This change requires a small adaptation to FomanticUI source code to generate the icon.css with duotone icons definitions (see discussion in #36 for details).

See: #36

testinfected committed:
Add support for FontAwesome Pro Duotone icons

See: #36

lubber-de commented on src/templates/icon.overrides.liquid:
```diff
@@ -1,3 +1,33 @@
 /*******************************
-         Theme Overrides
+Theme Overrides
```
The indentation was intentional, please keep it, it's the same for all override files

lubber-de commented on src/templates/icon.overrides.liquid:
```diff
@@ -1,3 +1,33 @@
+i.icon.duotone::after {
+  color: var(--fa-secondary-color, inherit);
+  opacity: var(--fa-secondary-opacity, 0.4);
+}
```
please add a newline

lubber-de commented on src/templates/icon.overrides.liquid:
```diff
@@ -1,3 +1,33 @@
+
+/* Step 3: Set the default opacity levels and colors for each layer */
+i.icon.duotone::before {
+  color: var(--fa-primary-color, inherit);
```
How does this work in FUI style like `green icon`? How to supply the second color?
There is something like `green-orange` classnaming logic needed instead of the fixed css variables.
Instead selectors like 
```css
/* Make sure to respect the `variation.variables` and `colors.less` file values when generating this*/
i.icon[class^="yellow-"]::before {
  color: yellow;
}
i.icon[class^="-green"]::after {
  color: green;
}
```
would be needed.

This works out as

```html
<i class="yellow-green camera-duotone icon"></i>
```

As duotone is pro plan only, those could be added to the override file as well (as it wont be part of fui core)
For the same reason, i would put the generation inside the overrides file as well.

```less
/* Yes, this _could_ be supported inside the existing generateIcons Funtion.... :) */

.generateSecondaryIcons(@map, @fontFamily) {
  each(@map,{
    @escapedKey: replace(@key,'^([0-9])','\3$1 ');
    @normalizedKey: replace(@escapedKey,'_','.','g');
    i.icon.@{normalizedKey}::after {
      content: "@{value}";
      & when not (@fontFamily = false) {
        font-family: @fontFamily;
      }
    }
  });
};

.generateIcons(@icon-duotone-map,false);
.generateSecondaryIcons(@icon-duotone-secondary-map,false);
.generateIcons(@icon-duotone-aliases-map,false);
.generateSecondaryIcons(@icon-duotone-secondary-aliases-map,false);
```

testinfected replied:
Ok so IIUC we could use class names like green-orange to specify the two icons colours. And we want to support all color variables right?

Concerning the generation I noticed that Thin, which are pro icons too, are not in override file. Shouldn't they be together?

lubber-de replied:
> Ok so IIUC we could use class names like green-orange to specify the two icons colours. And we want to support all color variables right?

Yes 

> Concerning the generation I noticed that Thin, which are pro icons too, are not in override file. Shouldn't they be together?

I agree, this is inconsistent...alright, lets move this part to the FUI core in a separate PR for FUI. As long as the default duotone maps are empty , nothing is being generated just as for thin. Good point.

testinfected replied:
`i.icon[class^="yellow-"]::before` will only match if the color is the first class of the icon. I suggest to change it to `i.icon[class~="yellow-"]::before` to match anywhere in the class list. What do you think?

lubber-de replied:
> `i.icon[class^="yellow-"]::before` will only match if the color is the first class of the icon. I suggest to change it to `i.icon[class~="yellow-"]::before` to match anywhere in the class list. What do you think?

👍🏼 Yes, my bad, that was my intention, so ~= is the way to go

testinfected replied:
Ah damn hyphen does not work with [attr~=value]. Any suggestion?

testinfected replied:
`black-orange` is a single word so [class~='black-'] does not match

testinfected replied:
We could define all combinations if done in a less template and use "black orange" instead

lubber-de replied:
[class*="yellow-"] works
https://jsfiddle.net/lubber/Ltzouv8k/5/

testinfected replied:
Of course, damn!

lubber-de commented on src/templates/icon.overrides.liquid:
```diff
@@ -1,3 +1,33 @@
+i.icon.duotone::before {
+  color: var(--fa-primary-color, inherit);
+  opacity: 1;
+  opacity: var(--fa-primary-opacity, 1.0);
```
Opacity line is doubled and should be taken from a separate LESS variable like `@duotonePrimaryOpacity` and `@duotoneSecondaryOpacity` instead

lubber-de commented on src/templates/icon.overrides.liquid:
```diff
@@ -1,3 +1,33 @@
+}
+
+i.icon.duotone::after {
+  color: var(--fa-secondary-color, inherit);
```
Same here. See above regarding color

lubber-de commented on src/templates/icon.overrides.liquid:
```diff
@@ -1,3 +1,33 @@
 *******************************/
+
+/* Step 1: Common Duotone Properties: All required to make icons render reliably */
+i.icon.duotone {
```
As all duotone icon names have a suffix, we could get rid of the additional classname (so "camera-duotone icon" is sufficient instead of "camera-duotone duotone icon"
```suggestion
i.icon[class^="-duotone"] {
```

testinfected replied:
I don't get this one. When I use duotone icons, I use `camera duotone icon` and not `camera-duotone duotone icon`

lubber-de replied:
> I don't get this one. When I use duotone icons, I use `camera duotone icon` and not `camera-duotone duotone icon`

You added `-duotone` to each duotone icon as name
![image](https://user-images.githubusercontent.com/18379884/199983100-ed2987e7-87c7-40f9-9945-b9e382c0ed8e.png)

The getClassName function takes this name and creates classnames out of this. As a result `camera-duotone` is being generated. If we don't do this, then , yes, a separate `duotone` class makes sense

testinfected replied:
Oh I see. You can guess I hacked duotone support without a deep understanding. I see that outline and thin icons define both the `-thin` and `.thin` way of using icons, and basically I copied that. Personnally I use `camera duotone` and not `camera-duotone`. I can remove the `-duotone` but it will make it different from how thin and outline icons work.

lubber-de replied:
I looked at the code again....infact it will NOT generate `-duotone`.. Sorry, it will, just as for outline and thin, generate a separate classname....all good, i will revoke my related review comments

lubber-de commented on src/templates/icon.overrides.liquid:
```diff
@@ -1,3 +1,33 @@
+}
+
+/* Step 2: Set the font-family and font-weight for this style */
+i.icon.duotone::before,
+i.icon.duotone::after {
```
Same here
```suggestion
i.icon[class^="duotone"::before,
i.icon[class^="duotone"::after {
```

lubber-de commented on src/templates/icon.overrides.liquid:
```diff
@@ -1,3 +1,33 @@
+  opacity: var(--fa-primary-opacity, 1.0);
+}
+
+i.icon.duotone::after {
```
Same here regarding selector
```suggestion
i.icon[class^="duotone"]::after {
```

lubber-de commented on src/templates/icon.overrides.liquid:
```diff
@@ -1,3 +1,33 @@
+
+/* Step 3: Set the default opacity levels and colors for each layer */
+i.icon.duotone::before {
+  color: var(--fa-primary-color, inherit);
```
We should use a LESS variable for defaults instead like `@duotonePrimaryColor`

lubber-de commented on src/templates/icon.overrides.liquid:
```diff
@@ -1,3 +1,33 @@
+
+i.icon.duotone::after {
+  color: var(--fa-secondary-color, inherit);
+  opacity: var(--fa-secondary-opacity, 0.4);
```
Same here regarding LESS variable for default secondary opacity

lubber-de reviewed (changes requested):
Please see my comments :)

testinfected commented:
Oh in depth review, I love it. It helps me better understand how to work with Fomantic source code. See my questions inline.

testinfected committed:
Add definitions for the Fomantic color palette

Add definitions for the various Fomantic colors, introduce variables
and rework/simplify positioning of icon layers.

See: #36

testinfected committed:
Match duotone color classes anywhere in the class names

See: #36

lubber-de commented on src/templates/icon.overrides.liquid:
```diff
@@ -1,3 +1,146 @@
+
+/* Primary icon color */
+i.icon.duotone[class~="red-"]::before {
+  color: red;
```
We should use the colors from the FUI color palette instead of the default browser constants which do not match the fui colors. 
So all color selectors have to be generated out of the colors.less file colors-map.
Let's do this inside the FUI code and move it out of this PR .
Should i do this in the FUI Repo or would you like to do it?

testinfected replied:
Ok help me out here :-)

testinfected replied:
Does that mean that I should remove all color definitions from icon.overrides.liquid?

lubber-de replied:
> Does that mean that I should remove all color definitions from icon.overrides.liquid?

Yes, this script should concentrate on creating the icons themselves

lubber-de commented on src/templates/icon.variables.liquid:
```diff
@@ -211,3 +248,10 @@ Icons are order A-Z in their group, Solid, Outline, Thin (Pro only) and Brand
 @massive: 8em;
+
+/* Duotone specifics */
+@duotonePrimaryColor: var(--fa-primary-color, inherit);
```
This wont work unless one declares --fa-primary-color somewhere in their code.
What i meant was to set a default color (hex value) as default instead of a css variable (and --fa-primary-color is from the original FA lib which isnt defined in FUI....and..(even if not really relevant anymore) the var() doesnt work in IE/legacy Edge

testinfected replied:
Ok so what should that default color be to fit with FUI?

lubber-de replied:
> Ok so what should that default color be to fit with FUI?

inherit should be sufficient i believe

testinfected replied:
Alright, I will make the changes over the week-end. Thanks for you help

lubber-de commented on src/templates/icon.variables.liquid:
```diff
@@ -211,3 +248,10 @@ Icons are order A-Z in their group, Solid, Outline, Thin (Pro only) and Brand
+@duotonePrimaryColor: var(--fa-primary-color, inherit);
+@duotoneSecondaryColor: var(--fa-secondary-color, inherit);
+
+@duotonePrimaryOpacity: var(--fa-primary-opacity, 1.0);
+@duotoneSecondaryOpacity: var(--fa-secondary-opacity, 0.4);
```
Same here, let's simply do
```suggestion
@duotonePrimaryOpacity: 1;
@duotoneSecondaryOpacity: 0.4;
```

lubber-de reviewed (changes requested):
I commented again 😉

lubber-de reviewed (changes requested):
I prepared a PR in FUI and adopted everything you put in the .overrides and moved them into the less-file instead
https://github.com/fomantic/Fomantic-UI/pull/2543

So, you can revert any changes inside the `icon.overrives.liquid` file in your PR

testinfected commented:
Fantastic, will do. ThanksEnvoyé de mon iPhoneLe 4 nov. 2022 à 18:40, lubber-de ***@***.***> a écrit :﻿
@lubber-de requested changes on this pull request.

I prepared a PR in FUI and adopted everything you put in the .overrides and moved them into the less-file instead
fomantic/Fomantic-UI#2543
So, you can revert any changes inside the icon.overrives.liquid file in your PR

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

testinfected committed:
Remove overrides now that they've been integrated in FUI source

See: #36

lubber-de reviewed (approved):
LGTM
